elastic/beats
# Use `fingerprint` file identity by default and migrate file state from `native` or `path`

## Proposed commit message


This commit changes the default `file_identity` from `native` to
`fingerprint`, any previous state from `native` (or `path`) is
automatically migrated to `fingerprint` whe Filestream is starting.

The Filestream input has always had the [ability to update file identifiers](https://github.com/elastic/beats/blob/4278366ab03221e8b62183dc06f9505f6ccc5209/filebeat/input/filestream/prospector.go#L104-L122),
however it never worked as expected, leading to full data duplication
when changing the file identity. This commit fixes it to allow
changing the file identity from `native` (inode + device ID) and
`path` to `fingerprint` without any data duplication.


## Checklist


## Disruptive User Impact
Because the `fingerprint` is the new default file identity, files are now only ingested when they reach at least 1024 bytes. The old default behaviour can be enabled by setting the file identity to `native` and disabling the fingerprint in the scanner.
```yaml
filebeat.inputs:
  - type: filestream
    id: "8.x-default-behaviour"
    paths:
      - /tmp/flog.log
    file_identity.native: ~
    prospector:
      scanner:
        fingerprint.enabled: false
```

## Author's Checklist
 - [x] Test with dynamic config reload
 - [x] Test with Kubernetes
 - [x] Test with Elastic-Agent
 - [x] Fix all the tests that break with the new behaviour
 - [x] Investigate which integration tests are going to break in the Elastic-Agent repo

Regarding the Elastic-Agent integration tests, most tests actually use the `log` input because when they were written, Filestream was not available as an integration package. The very few other test that use Filestrem either generate a log file large enough or are skipped as flaky.

## How to test this PR locally

1. Create a log file with at least a few log lines and more than 1kb (e.g: `/tmp/flog.log`, 15 log lines), you can use [`flog`](https://github.com/mingrammer/flog) with Docker:
    ```
    docker run -it --rm mingrammer/flog -n 15 > /tmp/flog.log
    ```
3. Start Filebeat with the following configuration

    <details><summary>filebeat.yml (native)</summary>
    <p>

    ```yaml
    filebeat.inputs:
      - type: filestream
        id: "test-migrate-ID"
        paths:
          - /tmp/flog.log
        file_identity.native: ~
        prospector:
          scanner:
            check_interval: 0.1s
            fingerprint.enabled: false
    
    queue.mem:
      flush.timeout: 0s
    
    output.file:
      path: ${path.home}
      filename: "output-file"
      rotate_on_startup: false
    
    logging:
      level: debug
      selectors:
        - input
        - input.filestream
        - input.filestream.prospector
      metrics:
        enabled: false
    ```

    </p>
    </details> 

4. Wait until the file is fully ingested (wait for `End of file reached: /tmp/flog.log; Backoff now.` in the logs)
5. Ensure all events have been published to the output (`wc -l ./output-file*` should return 15)
6. Stop Filebeat
7. Change the [file identity](https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-input-filestream.html#_file_identity_2) to `fingerprint`. It's the new default, hence it's not explicitly set.
    <details><summary>filebeat.yml (fingerprint)</summary>
    <p>

    ```yaml
    filebeat.inputs:
      - type: filestream
        id: "test-migrate-ID"
        paths:
          - /tmp/flog.log
        prospector:
          scanner:
            check_interval: 0.1s

    queue.mem:
      flush.timeout: 0s

    output.file:
      path: ${path.home}
      filename: "output-file"
      rotate_on_startup: false

    logging:
      level: debug
      selectors:
        - input
        - input.filestream
        - input.filestream.prospector
      metrics:
        enabled: false
    ```

    </p>
    </details> 
8. Start Filebeat
9. Wait until the Filebeat "finds the end of the file" (wait for `End of file reached: /tmp/flog.log; Backoff now.` in the logs)
10. Ensure no extra event was published ((`wc -l ./output-file*` should still return 15)
11. Add 10 more lines to the file:
    ```
    docker run -it --rm mingrammer/flog -n 10 >> /tmp/flog.log
    ```
13. Wait until the new lines are ingested (wait for `End of file reached: /tmp/flog.log; Backoff now.` in the logs)
14. Ensure all events have been published to the output with no duplication (`wc -l ./output-file*` should return 25)

## Related issues

- Closes https://github.com/elastic/beats/issues/40197

## Use cases
Dealing with identity reuse (e.g: inode reuse) without facing re-ingestion of data with Filestream input

~~## Screenshots~~

## Logs
